Convert JoinmarketQt to PySide2, qt5reactor

AdamISZ opened the pull request:
As per title.

Seems to be working in most functions, as I've tested in Python2 and 3 on regtest, except a rather important bug: it seems that the tumbler repeatably, but non-deterministically (haven't yet seen a pattern, happens "somewhere" in the run) segfaults the Qt app. Doesn't *seem* to happen in Python3, but obviously can't guarantee it.

I'll do a few tests on mainnet now to establish whether there are any further bugs in wallet handling there (since it uses some different code etc.).

For this unfortunate segfault issue, it's going to be hard; I can only wonder if it might have something to do with threading, as an earlier crash problem in Py2/PyQt4 ended up being to do with that; I found it could be fixed using pyqt4reactor and ensuring there was no threading used anywhere. But I could be totally off base there. Any suggestions on a good way to debug and find out the cause would be appreciated.

W.r.t. installation, for now I've just put `PySide2` into the jmclient setup.py; I've found `pip install PySide2` to be working fine on both Py2 and Py3, so I *guess* that's enough?

AdamISZ commented:
Oh, just realised I forgot that the direct pip(3) install of qt5reactor doesn't work anyway, this will make installation a bit complicated (or can just offload to user via doc), they need to use the sunu fork https://github.com/sunu/qt5reactor

undeath commented:
https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/86684c036a7e059dcf973d20eb271f793b114939/scripts/qtsupport.py#L342

the division operator needs to be changed to `//` here

jameshilliard commented on scripts/qtsupport.py:
```diff
@@ -20,11 +20,19 @@
+import sys
+is_py2 = sys.version[0] == '2'
+if is_py2:
+    import Queue as queue
```
This shouldn't be needed. Just do `import queue as Queue` which should work for both py2 and py3.

AdamISZ replied:
Thanks for the tip; realised it's not needed, see https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/248/commits/61e11cc5b27fac1b7c3198f466f159f4f681148b

AdamISZ commented:
Just to record what I experienced trying to test this on Python2:

Everything seemed to work fine (basically, patches needed here and there ofc) on mainnet until I started doing coinjoins - connecting to IRC is fine, but as the offers come in after `!orderbook` the Qt application reliably, but nondeterministically, crashes and segfaults. So here I'm basically repeating the report I made in the [first](https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/248#issue-237140160) comment, but more emphatically: on mainnet, it happens *every time*, not just occasionally.

I spent a long while, a couple of days at least, trying to debug this. There seemed to be two distinct ways the crash could occur, albeit closely connected. Both related to the `QSocketNotifier` class. To save myself writing I'll copy-paste what I said in IRC:

```
 <waxwing> suspicious how similar this is; even though a different error message, it's a failure on exactly the same line: https://github.com/Syncplay/syncplay/issues/147#issuecomment-336956403
<waxwing> this looks very relevant: https://bugreports.qt.io/browse/PYSIDE-629?focusedCommentId=395344&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel
<waxwing> especially because of py2.7 reference
<waxwing> given the segfaults described from here, it may well be the exact same bug https://bugreports.qt.io/browse/PYSIDE-629?focusedCommentId=407267&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-407267
<waxwing> been spending a lot of time trying to figure this out. no success really yet. gdb is giving me output very similar to this: https://pastebin.com/5bJ41ri9
<waxwing> which is coming out of the above thread.
<waxwing> the QSocketNotifier constructor seems to not be the only place it fails, i think when the 'activated' signal is triggered (of that notifier object), that may also be calling a segfault, but it's not so easy to figure out. what i can't seem to do is get to debug what is triggering/activating the signal, so, because its just segfaulting on receiving it, i'm in the dark.
<waxwing> the pattern of behaviour is: python2 only, pyside2 using qt5reactor, segfaults at some point after asknig for the orderbook on mainnet.
<waxwing> there seem to be lots of events happening at that point as you'd expect :) but at this point basically no clue which events are triggering the crash, or if that even matters, since from the above bug report discussion, it may well be the case that the underlying pyside2 code is just wrong (albeit, if i knew more i might be able to work round it)
```

Given the above (especially the pastebin from the thread) I'm tempted to believe that this is actually a real issue with PySide2 on Py2. So I've decided to abandon pursuing this further and, for now, having the Qt only compatible with Python3. I think there are very few users of the Qt app, and quite possibly none who would care about it only working on Python3, so not a sensible use of time (albeit I'd love to understand better what the bug is!).

AdamISZ commented:
I've been testing this a bit, and consider it ready to merge after squashing.

If anybody else intends to test it, please let me know.

AdamISZ commented:
Rebased and squashed. Will add some notes to README for install. As mentioned above, it'll be Py3 only, and the requirement should be `pip3 install PySide2` and `git clone https://github.com/sunu/qt5reactor; git checkout 58410aaead2185e9917ae9cac9c50fe7b70e4a60; cd <there>; python setup.py install`.

To explain again, the latest version from pip, which is the last release from that repo, is only PyQt5 compatible, not PySide2 compatible. So I don't think I see a better alternative (although of course it can be automated with a `--with-gui` flag in the installation, but that's kind of orthogonal to the question of how to deal with the fact that there isn't an official release with the needed functionality).
